temporal-sdk/src/main/java/io/temporal/worker/tuning/JVMSystemResourceInfo.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker.tuning;
+
+import com.sun.management.OperatingSystemMXBean;
+import io.temporal.common.Experimental;
+import java.lang.management.ManagementFactory;
+
+/** {@link SystemResourceInfo} implementation that uses JVM-specific APIs to get resource usage. */
+@Experimental
+public class JVMSystemResourceInfo implements SystemResourceInfo {
+  // As of relatively recent Java versions (including backports), this class will properly deal with
+  // containerized environments as well as running on bare metal.
+  // See https://bugs.openjdk.org/browse/JDK-8226575 for more details on which versions the fixes
+  // have been backported to.
+  OperatingSystemMXBean osBean = ManagementFactory.getPlatformMXBean(OperatingSystemMXBean.class);
+
+  @Override
+  @SuppressWarnings("deprecation") // deprecated APIs needed since replacements are for Java 14+
+  public double getCpuUsagePercent() {
```

Comment:
It looks like this and `getMemoryUsagePercent()` are called on every attempted slot reservation. Can you confirm that these JMX calls and Runtime calls are cheap and/or whether we should memoize for a short bit?

Reply:
So there is throttling at the slot supplier level, but yeah that could still potentially make for some rapid calls from different suppliers. I will investigate how expensive they are.